flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -0,0 +1,469 @@
+			//expected async error from close()
+
+			Assert.assertTrue(e.getMessage().contains("Error while sending value"));
+
+			//Final pending updates should be zero
+			Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+
+			//done
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	/**
+	 * Test ensures that an asyncError would be thrown on invoke() if previously message delivery failed.
+	 */
+	//TODO: should unitfy error handling logic in CassandraSinkBase
+	//Exception would have been thrown from invoke(), but asyncError was not set null, hence it was rethrown in close()
+	@Ignore
+	@Test
+	public void testAsyncErrorThrownOnInvoke() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+
+		casSinkFunc.open(new Configuration());
+
+		casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_FAILURE);
+		try {
+			casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_SUCCESS);
+		} catch (IOException e) {
+			//expected async error thrown from invoke()
+
+			//Final pending updates should be zero
+			Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+
+			casSinkFunc.close();
+
+			//done
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	/**
+	 * Test ensures that an asyncError would be thrown when checkpoint performs if previously message delivery failed.
+	 */
+	@Test
+	public void testAsyncErrorThrownOnCheckpoint() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness =
+			new OneInputStreamOperatorTestHarness<>(new StreamSink<>(casSinkFunc));
+
+		testHarness.open();
+
+		casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_FAILURE);
+
+		try {
+			testHarness.snapshot(123L, 123L);
+		} catch (Exception e) {
+			//expected async error from snapshotState()
+
+			Assert.assertTrue(e.getCause() instanceof IllegalStateException);
+			Assert.assertTrue(e.getCause().getMessage().contains("Failed to send data to Cassandra"));
+
+			//Final pending updates should be zero
+			Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+
+			casSinkFunc.close();
+
+			//done
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	///////////////////////
+	// Multi Thread Test
+	///////////////////////
+
+	/**
+	 * Test ensures that CassandraSinkBase would flush all in-flight message on close(), accompanied with concurrent
+	 * message delivery successfully via a thraedpool.
+	 */
+	@Test
+	public void testFlushOnPendingRecordsOnCloseWithSuccessfulMessage() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+
+		ExecutorService threadPool = Executors.newFixedThreadPool(MAX_THREAD_POOL_SIZE);
```

Comment:
For a given sink instance invoke() is only called by a single thread, so we're testing undefined behavior here by using multiple threads. (also applies to other tests)

Reply:
Okay. I can get rid of those multi-threading tests with mocks to mimic scenarios to ensure it will wait upon flushing all pending records before checkpoint was triggered.